app/Actions/AlbumAuthorisationProvider.php

```diff
@@ -246,51 +263,33 @@ public function applyReachabilityFilter(Builder $query): Builder
 	 *  - the root album is accessible by everybody
 	 *  - the built-in smart albums are accessible, if
 	 *     - the user is authenticated and is granted the right of uploading, or
-	 *     - the album is the album of recent photos and public by configuration, or
-	 *     - the album is the album of starred photos and public by configuration
+	 *     - the album is public
 	 *
-	 * @param string|null $albumID
+	 * @param AbstractAlbum|null $album
 	 *
 	 * @return bool
 	 */
-	public function isAccessibleByID(?string $albumID): bool
+	public function isAccessible(?AbstractAlbum $album): bool
 	{
-		// the admin may access everything, the root album may be accessed by everybody
-		if (AccessControl::is_admin() || empty($albumID)) {
+		if ($album === null || AccessControl::is_admin()) {
 			return true;
 		}
 
-		// Deal with built-in smart albums
-		if ($this->albumFactory->isBuiltInSmartAlbum($albumID)) {
-			return $this->isAuthorizedForSmartAlbum(
-				$this->albumFactory->createSmartAlbum($albumID)
-			);
-		}
-
-		// Use `applyAccessibilityFilter` to build a query, but don't hydrate
-		// a model
-		return $this->applyAccessibilityFilter(
-			BaseAlbumImpl::query()->where('base_albums.id', '=', $albumID)
-		)->count() !== 0;
-	}
-
-	public function isAccessible(Album $album): bool
-	{
-		if (AccessControl::is_admin()) {
-			return true;
-		}
-		if (!AccessControl::is_logged_in()) {
-			return
-				($album->is_public && $album->password === null) ||
-				($album->is_public && $this->isAlbumUnlocked($album->id));
-		} else {
-			$userID = AccessControl::id();
+		$userID = AccessControl::is_logged_in() ? AccessControl::id() : null;
 
+		if ($album instanceof BaseAlbum) {
 			return
 				($album->owner_id === $userID) ||
 				($album->is_public && $album->password === null) ||
-				($album->is_public && $this->isAlbumUnlocked($album->id)) ||
+				($album->is_public && $this->isUnlocked($album)) ||
 				($album->shared_with()->where('user_id', '=', $userID)->count());
+		} elseif ($album instanceof BaseSmartAlbum) {
+			return
+				($userID && AccessControl::can_upload()) ||
```

Comment:
can_upload() is already checking if user is logged in, so you don't need the `$userId` here.

Reply:
Actually, it is not identical and it already has driven me crazy. If no admin user is configured, i.e. if the password and the user name of the admin equals the empty string `'' `, then many parts of the application behave as if the admin user was logged in, in particular `is_logged_in` returns `true` (and so does `can_upload`) although `$userID` equals `null`.

IMHO, this is bad design. First of all, many methods are dangerous misnomers. If there is no user, `is_logged_in` and similar methods should return `false`. Moreover, the code should precisely distinguish between `is_admin` and `has_admin_privileges`. The former should return `true`, if and only if a proper admin user exists and the admin user is properly logged in, but `false` otherwise. The latter method might also return `true`, if there is no admin yet and thus any user has admin privileges. Unfortunately, the current code base is extremely bad with respect to this whole issue.

Long story short, for the moment I would not change this but leave it as is. In the long run, I am considering to completely replace the whole home-brewed `SessionFunctions` by the proper, built-in Laravel component. This would also allow to use `request->user()`, because the user object is injected into the request by Laravel middleware. At the moment this does not work. In this context, I would also clean up this whole mess around privileges and how they are named.

Reply:
Fixed in: https://github.com/LycheeOrg/Lychee/pull/1217/commits/da561cb361a0a4919cfa53f487e1e26ce3ab6981